src/ui/ui_i18n/README.md

```diff
@@ -0,0 +1,169 @@
+## I18n engine
+
+[React-intl](https://github.com/yahoo/react-intl) is built around
+[intl-messageformat](https://github.com/yahoo/intl-messageformat),
+so both react and angular frameworks will use the same engine and
+the same message syntax. The engine uses the ICU Message syntax
+and works for all CLDR languages which have pluralization rules
+defined. `intl-messageformat` package exposes `IntlMessageFormat`
+constructor. Messages are provided into the constructor as a
+string message, or a pre-parsed AST object.
+
+```
+import IntlMessageFormat from 'intl-messageformat';
+
+const msg = new IntlMessageFormat(message, locales, [formats]);
+```
+
+The string `message` is parsed, then stored internally in a
+compiled form that is optimized for the `format()` method to
+produce the formatted string for displaying to the user.
+
+```
+const output = msg.format(values);
+```
+
+Creating instances of `IntlMessageFormat` is an expensive operation.
+[Intl-format-cache](https://github.com/yahoo/intl-format-cache)
+library is simply to make it easier to create a cache of format
+instances of a particular type to aid in their reuse. Under the
+hood, this package creates a cache key based on the arguments passed
+to the memoized constructor.
+
+```
+import memoizeIntlConstructor from 'intl-format-cache';
+
+const getMessageFormat = memoizeIntlConstructor(IntlMessageFormat);
+```
+
+`Intl-messageformat` package assumes that the
+[Intl](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl)
+global object exists in the runtime. `Intl` is present in all modern
+browsers and Node.js > 0.10. In order to load `Intl MessageFormat`
+in Node.js we should simply `require()` this package (in Node.js,
+the data for all 200+ languages is loaded along with the library):
+```
+const IntlMessageFormat = require('intl-messageformat');
+```
+
+After that we are able to use this library in the same way as for browsers.
+
+## Angular wrapper
+
+Angular wrapper will have at least 3 entities: translation `service`,
+`directive` and `filter`. Both, the directive and the filter will use
+the translation service with `intl-messageformat` library under the hood.
+
+The translation `service` will provide the following methods (names can be changed):
+- `addMessages(messages: object, [locale: string])` - provides a way to register
+translations with the library
+- `getMessages()` - returns messages for the current language
+- `locale(locale: string)` - tells the library which language to use by given
+language key
+- `locale()` - returns the current locale 
+- `defaultLocale(locale: string)` - tells the library which language to fallback
+when missing translations
+- `defaultLocale()` - returns the default locale
+- `defineFormats(formats: object)` - supply a set of options to the underlying formatter
+- `translate({id: string, values: object, defaultMessage: string})` – translate message by id
+
+The translation `filter` will look like this:
```

Comment:
Hmm, so `translationId_placeholder` is a convention to mean the placeholder attribute on elements of id `translationId`? That seems sort of unexpected, but I'm not 100% opposed to it. I'd prefer a more distinct syntax personally, maybe `translationId[placeholder]` or something... something that wouldn't be a valid I'd otherwise, but then we're encoding HTML details into the translation keys...

Reply:
> Hmm, so translationId_placeholder is a convention to mean the placeholder attribute on elements of id translationId?

Yes, with `i18n-id="translationId"`.

> I'd prefer a more distinct syntax personally, maybe translationId[placeholder] or something... something that wouldn't be a valid id otherwise

Yeah, `translationId_placeholder` was just an example, if we want to implement this idea we'll think of a better naming convention :) Even if we put this idea aside, we'll likely need to express that label is a `placeholder` or `aria-label` or `tooltip` in its name anyway (unless we want to always add context comments in translation files that translators can use to get the context)

> but then we're encoding HTML details into the translation keys...

That's definitely a valid concern, I used the trick to persuade myself that this is "fine": it's not about __HTML attributes__, it's about __message attributes__ that are tied to a particular message and used in the same context and it just happened so that __HTML attributes__ are easily mapped to these abstract __message attributes__, it's arguable though :) In JSON it may look weird, I agree, it looks much better in FTL files (just as an example):
```
login-input = Predefined value
    .placeholder = example@email.com
    .aria-label = Login input value
    .title = Type your login email
```

Reply:
In general I like the idea, but I see some concerns with this approach:
- we need to think about how to pass different variables into different attributes
- it's not a standard syntax and it can confuse new developers

Reply:
It's possible, but we can get names collision if 2 translation keys will have the same variable names, so developers should keep in mind this fact. We can add extra checking for translation keys and show warnings for duplicating variable names in the same scope

Reply:
>  but we can get names collision if 2 translation keys will have the same variable names,

Sorry, not sure I'm following, why do you think this is a problem and what stops us from re-using the same value/argument in placeholder and title for the same element? They are used in the context of the same element, so sharing arguments doesn't sound like something odd to me.